pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -0,0 +1,139 @@
+	var allExpressions []string
+	allExpressions = append(allExpressions, validateString(we.Input)...)
+	for _, value := range we.Values {
+		allExpressions = append(allExpressions, validateString(value)...)
+	}
+	return allExpressions, len(allExpressions) != 0
+}
+
+func (we *WhenExpression) valuesContainsInput() bool {
+	for i := range we.Values {
+		if we.Values[i] == we.Input {
+			return true
+		}
+	}
+	return false
+}
+
+func (we *WhenExpression) isTrue() bool {
+	if we.Operator == selection.In {
+		return we.valuesContainsInput()
+	}
+	return !we.valuesContainsInput()
+}
+
+func (we *WhenExpression) hasVariable() bool {
+	if _, hasVariable := we.GetVarSubstitutionExpressionsForWhenExpression(); hasVariable {
+		return true
+	}
+	return false
+}
+
+func (we *WhenExpression) applyReplacements(replacements map[string]string) WhenExpression {
+	replacedInput := ApplyReplacements(we.Input, replacements)
+
+	var replacedValues []string
+	for _, val := range we.Values {
+		replacedValues = append(replacedValues, ApplyReplacements(val, replacements))
+	}
+
+	return WhenExpression{Input: replacedInput, Operator: we.Operator, Values: replacedValues}
+}
+
+// WhenExpressions is an array of When Expressions which are used to specify whether a Task should be executed or skipped.
+type WhenExpressions []WhenExpression
+
+// AreTrue indicates whether the When Expression are true.
+func (wes WhenExpressions) AreTrue() bool {
+	for _, we := range wes {
+		if isTrue := we.isTrue(); !isTrue {
+			return false
+		}
+	}
+	return true
+}
+
+// HaveVariables indicates whether When Expressions have variables in the Inputs or Values.
+func (wes WhenExpressions) HaveVariables() bool {
+	for _, we := range wes {
+		if hasVariable := we.hasVariable(); !hasVariable {
+			return false
+		}
+	}
+	return true
+}
+
+// ReplaceWhenExpressionsVariables interpolates variables in When Expressions.
+func (wes WhenExpressions) ReplaceWhenExpressionsVariables(replacements map[string]string) WhenExpressions {
+	var replaced []WhenExpression
+	for _, we := range wes {
+		replaced = append(replaced, we.applyReplacements(replacements))
+	}
+	return replaced
+}
+
+// WhenExpressionEvaluationResult is the result from evaluating a When Expression.
+type WhenExpressionEvaluationResult struct {
+	// Expression is the expression, made up of Input, Operator and Values.
+	Expression *WhenExpression `json:"expression,omitempty"`
+	// IsTrue is a boolean from evaluating the Expression.
+	IsTrue bool `json:"result,omitempty"`
+}
+
+// WhenExpressionsEvaluationStatus are the results from evaluating When Expressions.
+type WhenExpressionsEvaluationStatus struct {
+	// EvaluationResults are the results from evaluating When Expressions.
+	EvaluationResults []*WhenExpressionEvaluationResult `json:"evaluationResults,omitempty"`
+	// Executed is a boolean that represents an AND of all results from evaluating When Expressions.
+	Executed bool `json:"execute,omitempty"`
```

Comment:
are you certain we need Executed in addition to the IsTrue values in EvaluationResults?

i also think this might duplicate information from the Condition, which i assume will also contain information about the fact that this was skipped and why?

Reply:
We are storing all the when expressions specified and their results in [status.taskruns[].whenExpressionsStatus](https://gist.github.com/pritidesai/cbc7caadbd80f02745383acf524e3186#file-pull-3135-pipelinerun-json-L200) but do we really need to? It would be beneficial for debugging to see the resolved when expressions but at the same time makes the status too verbose. Instead, if we can store the resolved when expressions in the [status.pipelineSpec.tasks[].when](https://gist.github.com/pritidesai/cbc7caadbd80f02745383acf524e3186#file-pull-3135-pipelinerun-json-L125), we can avoid these structs and simplify the following function `GetWhenExpressionStatus`.

Reply:
`GetWhenExpressionStatus` then needs to iterate over when expressions and return false if any of the expression evaluates to false i.e. `!wes[i].isTrue()`.

Reply:
**This is separate from this PR**

I was having trouble seeing unresolved when expressions in `status.pipelineSpec.tasks[]` and was trying to troubleshoot and discovered we actually store the `pipelineSpec` as is even before resolving params and it's never updated for auditing purpose. But might be clearer if its updated once after resolving all params (and pipeline resources).

Reply:
In that case, it make sense to store resolved when expressions in the status 😢

Reply:
After reading #3140, we might want to reconsider storing `when` expressions in the status (not trying to be a blocker here, this can be updated in next cycle which I believe we should, thanks @jerop for creating an issue #3145 to track this) 🙏